### ISSUE TITLE
Iterator compilation crashes when known types aren't available

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#9463]
The test below fails, but it should report diagnostics instead for some types and degrade gracefully for other types. 
To repro, make sure the corlib is missing `IDisposable.Dispose` or `IEnumerator.MoveNext` or `IEnumerable<T>.GetEnumerator` or `System.Threading.Thread` as those types will be added to the shared mincorlib shortly.  
This problem also exists with the VB compiler.

On the C# side, the exception is thrown by `Debug.Assert(useSiteDiagnostics.IsNullOrEmpty());` in `SyntheticBoundNodeFactory.Return(BoundExpression expression)` at SyntheticBoundNodeFactory.cs(426,0).

From discussion with Tomas and Neal, the way to fix this for type dependencies declared in the language specification is to move the try/catch for missing predefined members around the call to the rewriter and emit a `throw null` instead. The try/catch is currently in `IteratorRewriter.GenerateMethodImplementations()`.

But the expected behavior for some types (like `Thread`) is for the compilation to succeed without diagnostic message, since this dependency is an implementation detail and not specified by the language specification. The implementation should degrade gracefully (somehow) if this type is missing.  
Update: per discussion with Jared, it's probably better to leave the Thread behavior as-is for now.

Note: the same problem probably exists with the async rewriter as well.

```C#
        public void IteratorRequireSpecialKnownTypesInCorLib()
        {
            // IDisposable.Dispose and IEnumerable.MoveNext

            var source =
@"
public class C
{
    public System.Collections.IEnumerable SomeNumbers() 
    {
        yield return 42;
    }
}";
            var comp = CreateCompilation(new[] { Parse(source) }, new[] { TestReferences.NetFx.Minimal.mincorlib });
            CompileAndVerify(comp);
        }
```

This was discovered while investigating #9167, which is about compiling async and iterator methods in VB while missing expected types.

**Comment by jcouv** [comment]
Commit c41d4f8300d78a1117721d42bcca1307789bbbfd addresses the iterator rewriter part of the problem. I'll start on the async rewriter.
